Mathlib/Data/Matrix/PEquiv.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+
+! This file was ported from Lean 3 source module data.matrix.pequiv
+! leanprover-community/mathlib commit 3e068ece210655b7b9a9477c3aff38a492400aa1
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Matrix.Basic
+import Mathlib.Data.PEquiv
+
+/-!
+# partial equivalences for matrices
+
+Using partial equivalences to represent matrices.
+This file introduces the function `PEquiv.toMatrix`, which returns a matrix containing ones and
+zeros. For any partial equivalence `f`, `f.toMatrix i j = 1 ↔ f i = some j`.
+
+The following important properties of this function are proved
+`toMatrix_trans : (f.trans g).toMatrix = f.toMatrix ⬝ g.toMatrix`
+`toMatrix_symm  : f.symm.toMatrix = f.toMatrixᵀ`
+`toMatrix_refl : (PEquiv.refl n).toMatrix = 1`
+`toMatrix_bot : ⊥.toMatrix = 0`
+
+This theory gives the matrix representation of projection linear maps, and their right inverses.
+For example, the matrix `(single (0 : Fin 1) (i : Fin n)).toMatrix` corresponds to the ith
+projection map from R^n to R.
+
+Any injective function `Fin m → Fin n` gives rise to a `PEquiv`, whose matrix is the projection
+map from R^m → R^n represented by the same function. The transpose of this matrix is the right
+inverse of this map, sending anything not in the image to zero.
+
+## notations
+
+This file uses the notation ` ⬝ ` for `Matrix.mul` and `ᵀ` for `Matrix.transpose`.
+-/
+
+
+namespace PEquiv
+
+open Matrix
+
+universe u v
+
+variable {k l m n : Type _}
+
+variable {α : Type v}
+
+open Matrix
+
+-- Porting note: added
+instance [DecidableEq n] (j : n) (o : Option n) : Decidable (j ∈ o) := by
+  rw [←Option.mem_toList]
+  infer_instance
+
+/-- `toMatrix` returns a matrix containing ones and zeros. `f.toMatrix i j` is `1` if
+  `f i = some j` and `0` otherwise -/
+def toMatrix [DecidableEq n] [Zero α] [One α] (f : m ≃. n) : Matrix m n α :=
+  of fun i j => if j ∈ f i then (1 : α) else 0
+#align pequiv.to_matrix PEquiv.toMatrix
+
+-- TODO: set as an equation lemma for `toMatrix`, see mathlib4#3024
+@[simp]
+theorem toMatrix_apply [DecidableEq n] [Zero α] [One α] (f : m ≃. n) (i j) :
+    toMatrix f i j = if j ∈ f i then (1 : α) else 0 :=
+  rfl
+#align pequiv.to_matrix_apply PEquiv.toMatrix_apply
```

Comment:
Is this something we need to do before merging?

Reply:
I don't think we can do this before #3024 lands. CC @eric-wieser

Reply:
Yes, we don't need to do this before merging. As long as we leave behind the references to 3024, then there is no rush to actually merge the PR. We can do a second pass over these comments later.